libtracepoint-control-cpp/src/TracepointSession.cpp

```diff
@@ -1021,221 +1065,38 @@ TracepointSession::SavePerfDataFile(
-        {
-            errno = 0;
-            tpi.m_bufferFiles[bufferIndex].reset(perf_event_open(pAttr, -1, bufferIndex, -1, PERF_FLAG_FD_CLOEXEC));
-            if (!tpi.m_bufferFiles[bufferIndex])
-            {
-                error = errno;
-                if (error == 0)
-                {
-                    error = ENODEV;
-                }
-
-                goto Error;
-            }
-        }
-
-        if (m_bufferLeaderFiles)
-        {
-            // Leader already exists. Add this event to the leader's mmaps.
-            error = IoctlForEachFile(tpi.m_bufferFiles.get(), m_bufferCount, PERF_EVENT_IOC_SET_OUTPUT, m_bufferLeaderFiles);
-            if (error)
-            {
-                goto Error;
-            }
-        }
-        else
-        {
-            auto const mmapSize = m_pageSize + m_bufferSize;
-            auto const prot = IsRealtime()
-                ? PROT_READ | PROT_WRITE
-                : PROT_READ;
-
-            // This is the first event. Make it the "leader".
-            for (uint32_t bufferIndex = 0; bufferIndex != m_bufferCount; bufferIndex += 1)
-            {
-                errno = 0;
-                auto cpuMap = mmap(nullptr, mmapSize, prot, MAP_SHARED, tpi.m_bufferFiles[bufferIndex].get(), 0);
-                if (MAP_FAILED == cpuMap)
-                {
-                    error = errno;
-
-                    // Clean up any mmaps that we opened.
-                    for (uint32_t bufferIndex2 = 0; bufferIndex2 != bufferIndex; bufferIndex2 += 1)
-                    {
-                        m_buffers[bufferIndex2].Mmap.reset();
-                    }
-
-                    if (error == 0)
-                    {
-                        error = ENODEV;
-                    }
-
-                    goto Error;
-                }
-
-                m_buffers[bufferIndex].Mmap.reset(cpuMap, mmapSize);
-            }
-        }
-
-        // Find the sample_ids for the new tracepoints.
-        for (; cIdsAdded != m_bufferCount; cIdsAdded += 1)
-        {
-            ReadFormat data;
-            error = tpi.Read(cIdsAdded, &data);
-            if (error != 0)
-            {
-                goto Error;
-            }
-
-            pIds[cIdsAdded] = data.id;
-
-            auto const added = m_tracepointInfoBySampleId.emplace(data.id, &tpi).second;
-            assert(added);
-            (void)added;
-        }
 
-        if (!m_bufferLeaderFiles)
-        {
-            m_bufferLeaderFiles = tpi.m_bufferFiles.get(); // Commit this event as the leader.
-        }
-
-        tpi.m_enableState = TracepointEnableState::Enabled;
-        error = 0;
-        goto Done;
-    }
-    catch (...)
+    auto const existingIt = m_tracepointInfoByCommonType.find(metadata.Id());
+    if (existingIt == m_tracepointInfoByCommonType.end())
     {
-        error = ENOMEM;
-        goto Error;
+        error = AddTracepoint(metadata, TracepointEnableState::Enabled);
```

Comment:
Note: Refactored the Add into a helper method. This was originally done because I needed to call it from two places. I have since reverted that part, but I like having it in a helper method so I kept that.